src/main/java/org/primefaces/model/TreeNodeChildren.java

```diff
@@ -71,13 +71,14 @@ else if ((index < 0) || (index > size())) {
             eraseParent(node);
             super.add(index, node);
             node.setParent(parent);
-            updateRowKeys(parent);
+            updateRowKeys(index, this.parent);
         }
     }
 
     @Override
     public boolean addAll(Collection<? extends TreeNode> collection) {
         Iterator<TreeNode> elements = (new ArrayList<TreeNode>(collection)).iterator();
+        int index = collection.size();
```

Comment:
Coud you move index declaration into changed if statement, it's not used anywhere else